zerver/webhooks/slack/view.py

```diff
@@ -1,60 +1,187 @@
+        }
+        for file in file_dict
+    ]
+    return files
+
+
+def get_message_body(text: str, sender: str, files: SlackFileListT) -> str:
+    # TODO: Reformat mentions in messages, currently its <@USER_ID>.
+    body = ZULIP_MESSAGE_TEMPLATE.format(sender=sender, text=text)
+    for file in files:
+        body += FILE_LINK_TEMPLATE.format(**file)
+    return body
+
+
+def is_challenge_handshake(payload: WildValue) -> bool:
+    return payload.get("type").tame(check_string) == "url_verification"
+
 
 @webhook_view("Slack", notify_bot_owner_on_invalid_json=False)
 @typed_endpoint
 def api_slack_webhook(
     request: HttpRequest,
     user_profile: UserProfile,
     *,
-    user_name: str,
-    text: str,
-    channel_name: str,
+    payload: JsonBodyPayload[WildValue],
     channels_map_to_topics: str | None = None,
 ) -> HttpResponse:
-    content = ZULIP_MESSAGE_TEMPLATE.format(message_sender=user_name, text=text)
-    topic_name = "Message from Slack"
-
-    if channels_map_to_topics is None:
+    # Handle initial URL verification handshake for Slack Events API.
+    if is_challenge_handshake(payload):
+        challenge = payload.get("challenge").tame(check_string)
         check_send_webhook_message(
             request,
             user_profile,
-            topic_name,
-            content,
-        )
-    elif channels_map_to_topics == VALID_OPTIONS["SHOULD_BE_MAPPED"]:
-        # If the webhook URL has a user_specified_topic,
-        # then this topic-channel mapping will not be used.
-        topic_name = f"channel: {channel_name}"
-        check_send_webhook_message(
-            request,
-            user_profile,
-            topic_name,
-            content,
-        )
-    elif channels_map_to_topics == VALID_OPTIONS["SHOULD_NOT_BE_MAPPED"]:
-        # This channel-channel mapping will be used even if
-        # there is a channel specified in the webhook URL.
-        check_send_webhook_message(
-            request,
-            user_profile,
-            topic_name,
-            content,
-            stream=channel_name,
+            "Integration events",
+            "Successfully verified webhook URL with Slack!",
         )
+        return json_success(request=request, data={"challenge": challenge})
+
+    # Prevent any Zulip messages sent through the Slack Bridge from looping
+    # back here.
+    if is_zulip_slack_bridge_bot_message(payload):
+        return json_success(request)
+
+    event_dict = payload.get("event", {})
+    event_type = event_dict.get("type").tame(check_string)
+
+    if event_type == "message":
+        raw_files = event_dict.get("files")
+        files = convert_raw_file_data(raw_files) if raw_files else []
+
+        raw_text = event_dict.get("text", "").tame(check_string)
+        text = convert_to_zulip_markdown(raw_text) if raw_text else ""
+
+        sender = event_dict.get("user").tame(check_none_or(check_string))
+        if sender is None:
```

Comment:
Added a logic to anticipate Slack's integration bot messages. we can add support for this once #31162 is merged, there are a couple of commits that decouple relevant functions from Slack incoming integration to handle this kind of message

Reply:
added `message_from_slack_integration_bot.json` to document these type of messages

Reply:
Good catch! Updates look good.